File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/digitalcash/Input.java
package com.github.dedis.popstellar.model.network.method.message.data.digitalcash;

import com.google.gson.annotations.SerializedName;

import java.util.Objects;

/** Object representing a transaction output to use as an input for this transaction */
public final class Input {

  @SerializedName("tx_out_hash")
  private final String txOutHash; // Previous (to-be-used) transaction hash

  @SerializedName("tx_out_index")
  private final int txOutIndex; // index of the previous to-be-used transaction

  @SerializedName("script")
  private ScriptInput script; // The script describing the unlock mechanism

  /**
   * @param txOutHash Previous (to-be-used) transaction hash
   * @param txOutIndex index of the previous to-be-used transaction
   * @param script The script describing the unlock mechanism
   */
  public Input(String txOutHash, int txOutIndex, ScriptInput script) {
    this.script = script;
    this.txOutHash = txOutHash;
    this.txOutIndex = txOutIndex;
  }

  public String getTxOutHash() {
    return txOutHash;
  }

  public int getTxOutIndex() {
    return txOutIndex;
  }

  public ScriptInput getScript() {
    return script;
  }

  @Override
  public String toString() {
    return "input{"
        + "tx_out_hash='"
        + txOutHash
        + '\''
        + ", tx_out_index="
        + txOutIndex
        + ", script="
        + script
        + '}';
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    Input txIn = (Input) o;
    return txOutIndex == txIn.txOutIndex
        && txOutHash.equals(txIn.txOutHash)
        && script.equals(txIn.script);
  }

  @Override
  public int hashCode() {
    return Objects.hash(getTxOutHash(), getTxOutIndex(), getScript());
  }
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/digitalcash/Input.kt
package com.github.dedis.popstellar.model.network.method.message.data.digitalcash

import com.github.dedis.popstellar.model.Immutable
import com.google.gson.annotations.SerializedName
import java.util.Objects

/** Object representing a transaction output to use as an input for this transaction */
@Immutable
class Input
/**
 * @param txOutHash Previous (to-be-used) transaction hash
 * @param txOutIndex index of the previous to-be-used transaction
 * @param script The script describing the unlock mechanism
 */
(
    // Previous (to-be-used) transaction hash
    @field:SerializedName("tx_out_hash")
    val txOutHash: String, // index of the previous to-be-used transaction
    @field:SerializedName("tx_out_index")
    val txOutIndex: Int, // The script describing the unlock mechanism
    @field:SerializedName("script") val script: ScriptInput
) {

  override fun toString(): String {
    return "input{tx_out_hash='$txOutHash', tx_out_index=$txOutIndex, script=$script}"
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) {
      return true
    }
    if (other == null || javaClass != other.javaClass) {
      return false
    }
    val txIn = other as Input
    return txOutIndex == txIn.txOutIndex && txOutHash == txIn.txOutHash && script == txIn.script
  }

  override fun hashCode(): Int {
    return Objects.hash(txOutHash, txOutIndex, script)
  }
}
